File: solutions/problem_0347_top_k_frequent_elements/solution.cpp
/*
 * Problem 347: Top K Frequent Elements
 * ===================================
 * Difficulty: Medium
 * Tags: Array, Hash Table, Divide and Conquer, Sorting, Heap (Priority Queue), Bucket Sort, Counting, Quickselect
 * Pattern: Hash Map / Heap
 *
 * Time Complexity:  O(n log k)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Quickselect - O(n) average time
        int k = k;
        nth_element(nums.begin(), nums.begin() + nums.size() - k, nums.end());
        return nums[nums.size() - k];
    }
};

File: solutions/problem_0347_top_k_frequent_elements/solution.py
"""
Problem 347: Top K Frequent Elements
===================================
Difficulty: Medium
Tags: Array, Hash Table, Divide and Conquer, Sorting, Heap (Priority Queue), Bucket Sort, Counting, Quickselect
Pattern: Hash Map / Heap

Time Complexity:  O(n log k)
Space Complexity: O(n)
"""

from collections import Counter

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        count = Counter(nums)
        return [x for x, _ in count.most_common(k)]
